app-exp-dma.tex

```diff
@@ -0,0 +1,157 @@
+\emph{Descriptors} are small data structures typically comprising the address of a data buffer, the size of the transaction (eg number of bytes outstanding to send) and other metadata.
+For example, how to increment addresses (allowing 2D/3D-style accesses) and what to do next after this buffer is completed.
+Descriptors can be used in roughly two ways.
+First is the fetching of descriptors from memory, either main memory or a more local memory.
+Second is the streaming of descriptors on another port, for example from a separate descriptor fetching unit (as in the LogiCORE AXI DMA above, and also the Altera Modular Scatter-Gather DMA~\cite{altera-embeddedperipheralsug}).
+In-memory descriptors can often be chained or used in arrays in order to form \emph{scatter-gather lists} or \emph{ring buffers} to give a higher order structure to groups of transactions.
+
+\subsubsection{Bus-mastering DMA}
+In this case a device is able to make its own memory read and write transactions, independent of any system DMA controller.
+This is common for application-class devices such as PCI Express (PCIe) endpoints, including network cards, NVMe and AHCI storage, GPUs, NPUs and so on.
+In many cases a CHERI system cannot introspect into the device in order to modify its address behavior or to add CHERI support to its internal DMA infrastructure.
+Additionally, such devices may not be known at design-time any may change dynamically, thanks to hotpluggable PCIe cards and Thunderbolt devices.
+
+In order to safely constrain such a device, we must interpose something between the device and the memory subsystem.
+This is necessary in order to both limit the damage the device can do, and to protect the integrity of a CHERI system.
+Current systems use an IOMMU (also called a System MMU) to apply a page-based protection and translation model.
+This has significant performance implications due to the highly dynamic behavior of complex devices.
+It may be desirable to augment or replace the IOMMU with some other structure which instead uses the capability model, along the same lines of using CHERI for protection and the MMU for translation.
+Examples of such schemes are given in Cheng and Markettos~\cite{DBLP:conf/micro/MarkettosBBNMW20, cheng2025-capchecker}.
+
+Where possible, a preferable design point is to integrated CHERI support into a bus-mastering device such that the hardware can be trusted to manipulate capabilities correctly.
+In this instance we are able to run untrusted I/O software stack safely on trustworthy CHERI-enabled hardware.
+An example of this is a CHERI-enabled GPU~\cite{UCAM-CL-TR-997,naylor26-simtight}, which trustworthily shares capabilities with a CHERI-enabled CPU core.
+
+\subsection{Composing CHERI and DMA}
+DMA controllers operate outside the world of CPU-derived capabilities.
+The question arises how to compose them.
+Below we describe some different levels of utilizing capabilities with DMA.
+
+\subsubsection{Level 0: DMA clears tags}
```

Comment:
This has implications on the software driver model, too: a completely CHERI-oblivious DMA engine grants the authority to access memory given only an integer address, even if it cannot forge capabilities capturing that authority.  As such, access to such DMA engines must be carefully controlled and restricted to TCB components.